crates/sqllib/src/dec.rs

```diff
@@ -0,0 +1,923 @@
+//! Operations on Decimal values
+
+use crate::{
+    cast_to_SqlDecimal_s, some_existing_operator, some_function2, some_operator,
+    some_polymorphic_function1, some_polymorphic_function2, unwrap_cast, HasOne, HasZero,
+    SqlString,
+};
+use dbsp::algebra::{MulByRef, OptionWeightType, F64};
+use dec::{Context, Decimal, ParseDecimalError, Rounding, TryFromDecimalError};
+use feldera_types::deserialize_without_context;
+use feldera_types::serde_with_context::{
+    serde_config::DecimalFormat, SerializeWithContext, SqlSerdeConfig,
+};
+use num::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, Zero};
+use rkyv::{string::StringResolver, DeserializeUnsized, Fallible, SerializeUnsized};
+use serde::de::Error as DeserError;
+use serde::ser::Error as SerError;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde_json::Number;
+use size_of::SizeOf;
+use std::{
+    fmt,
+    hash::{Hash, Hasher},
+    ops::{Add, AddAssign, Div, Mul, Neg, Sub},
+    sync::LazyLock,
+};
+
+const MAX_DIGITS: usize = 13;
+const DECIMAL_MAX_PRECISION: usize = 3 * MAX_DIGITS;
+
+pub type LargeDecimal = Decimal<MAX_DIGITS>;
+
+#[derive(Copy, Debug, Default, Clone)]
+pub struct SqlDecimal {
+    value: LargeDecimal,
+}
+
+pub(crate) type DecimalContext = Context<LargeDecimal>;
+
+#[doc(hidden)]
+pub(crate) fn is_error(context: &DecimalContext) -> bool {
+    let status = context.status();
+    status.insufficient_storage()
+        || status.overflow()
+        || status.invalid_operation()
+        || status.division_by_zero()
+        || status.division_undefined()
+        || status.division_impossible()
+        || status.underflow()
+        || status.division_by_zero()
+        || status.division_undefined()
+}
+
+/// Prototype clean context
+#[doc(hidden)]
+pub static CONTEXT_PROTO: LazyLock<DecimalContext> = LazyLock::new(|| {
+    let mut cx = DecimalContext::default();
+    // Default rounding mode
+    cx.set_rounding(Rounding::Down);
+    // This is 38, conveniently matching the max precision supported by severall SQL dialects
+    cx.set_max_exponent(DECIMAL_MAX_PRECISION as isize - 1)
+        .unwrap();
+    cx.set_min_exponent(-(DECIMAL_MAX_PRECISION as isize))
+        .unwrap();
+    cx
+});
```

Comment:
I see that the global Context is cloned every time it is used. It's a 28-byte struct so that could add up eventually. I wonder whether making it thread-local and then using it directly instead of cloning it would be better?

However, it's probably in the noise for overall performance, so I doubt it's worth optimizing yet.

Reply:
the context is mut for almost all operations, so there is no choice but to clone it.

Reply:
On a second thought, this may not be the right approach at all. Cloning the context and modifying it cannot be better than creating a fresh context every time. The way a pre-allocated context could (in theory) be useful is if we allowed each operation to modify it and then undo the changes.